<a href="https://colab.research.google.com/github/jerlinvillalobos/Laboratorios/blob/main/TP1/Laboratorio_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://github.com/randall-romero/EC4301/blob/master/Laboratorios/01--Introducci%C3%B3n-a-Python/figures/escuela-de-economia.png?raw=1" width="260" height="60">
<center>
    <b>EC4300 MICROECONOMETRÍA</b><br>
    <b>Asistente:  Jerlin Villalobos</b>
<br><br>
<b>Laboratorio 5:</b>
<br>    
<div style="font-size:250%;color:white; background-color: #0064b0;">Introducción a Python</div>
<div style="font-size:175%;color:white; background-color: #0064b0;">Parte 6: Módulos </div>    

</center>
<br><br>
<p style="font-size:120%;">Este cuaderno toma como fuente los cuadernos de Jupyter desarrollados por Phd. Randall Romero. Se toma el
    <a href="https://colab.research.google.com/github/randall-romero/EC4301/blob/master/Laboratorios/01--Introducci%C3%B3n-a-Python/Python-06--M%C3%B3dulos.ipynb"> Lab de python 06</a>



<i> Última actualización del cuaderno original: 2020-Ago-18
    <br>
    Creado: Agosto 3, 2024
</i>

<hr>

# Importando módulos

* Las definiciones de funciones en Python en uno o más archivos separados para facilitar su manteniniento y para permitir usarlos en varios programas sin copiar las definiciones en cada uno.

* Cada archivo que almacena definciones de funciones se llama “módulo” y el nombre del módulo es igual al del archivo pero sin la extensión “.py”.

* Las funciones almacenadas en un módulo están disponibles para un programa usando la palabra clave ```import``` seguida del nombre del módulo.

* Aunque no es esencial, se acostumbra poner las instrucciones ```import``` al inicio del programa.

* Las funciones importadas pueden usarse llamando su nombre como un "punto-sufijo" luego del nombre del módulo. Por ejemplo, una función `sqrt` de un módulo importado `numpy` puede llamarse con `numpy.sqrt()`

## Algunos paquetes (módulos) muy útiles

En nuestro curso, los siguientes paquetes (= colecciones de módulos) serán muy útiles
* `numpy` Paquete base para arreglos N-dimensional. Operaciones matemáticas, especialmente álgebra lineal
* `matplotlib` gráficos 2D
* `pandas` estructuras para almacenar y analizar datos
* `scipy` librería fundamental para computación científica
* `bccr` ofrece funciones para descargar datos del Banco Central de Costa Rica
* `macrodemos` contiene demos de conceptos macroeconométricos, por ejemplo los modelos ARMA
* `compecon` Para resolver modelos de economía computacional

### Algunos ejemplos:

Para importar `numpy`

In [ ]:
import numpy
numpy.sqrt(9)

3.0

Mismo ejemplo, pero dándolo un “alias” al módulo

In [ ]:
import numpy as np
np.sqrt(9)

3.0

Mismo ejemplo, pero importando solo la función `sqrt`

In [ ]:
from numpy import sqrt
sqrt(9)

3.0

## ¿Por qué trabajar con módulos?

* Una ventaja de organizar el código en módulos y paquetes es evitar desordenar el espacio de nombres.
* Los módulos permites tener funciones del mismo nombre en espacios de nombre separados, obligándonos a ser explícitos acerca de cuál es la que usamos.

Por ejemplo, tanto `math` como `numpy` tienen una función `cos` para computar el coseno, pero su implementación es muy distinta.

### Con `numpy`:

$\pi$

In [ ]:
π = np.pi

In [ ]:
np.cos(π)

-1.0

In [ ]:
import numpy as np
print(np.cos(0))
print(np.cos([0,1, np.pi]))

1.0
[ 1.          0.54030231 -1.        ]


### Con `math`:

In [ ]:
import math
print(math.cos(0))
print(math.cos([0,1, np.pi]))

1.0


TypeError: must be real number, not list

El error se produce porque la función math.cos de la biblioteca math no acepta listas o arrays; solo acepta números individuales. La función math.cos está diseñada para trabajar con un solo valor numérico a la vez, no con arrays o listas

In [ ]:
#Una forma de usar math y que funcione
import math
import numpy as np

print(math.cos(0))
print([math.cos(x) for x in [0, 1, np.pi]])

1.0
[1.0, 0.5403023058681398, -1.0]


## Iteración más rápida

In [ ]:
nrep = 12000
values = list(range(nrep))   #range(nrep) genera una secuencia de números enteros desde 0 hasta 11,999 (es decir, 12,000 números en total).
                             #list(range(nrep)) convierte esta secuencia en una lista de números enteros.

In [ ]:
%%timeit
#mide el tiempo en ejecución del código
option0 = np.empty_like(values) #para almacenar los datos

for i, x in enumerate(values):
    option0[i] = math.cos(x)

4.26 ms ± 117 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


4.26 ms: El tiempo promedio que tarda en ejecutar una iteración del código en milisegundos.

± 117 µs: La desviación estándar, que muestra cuánto varía el tiempo de ejecución entre las distintas iteraciones. Aquí, 117 microsegundos es la variación promedio en el tiempo de ejecución.

per loop: Esto significa que el tiempo es para cada ejecución del bucle de código.

(mean ± std. dev. of 7 runs, 100 loops each): El tiempo mostrado es el promedio de 7 ejecuciones completas de %%timeit, donde cada ejecución consistió en 100 repeticiones del código para calcular el tiempo de ejecución más preciso.

In [ ]:
%%timeit
option1 = list()   #crear una lista vacía
for x in values:
    option1.append(math.cos(x))    #Para cada valor x, calcula el coseno usando math.cos(x) y añade el resultado a la lista option1

2.33 ms ± 78.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [ ]:
%%timeit
option2 = [math.cos(x) for x in values]

2.17 ms ± 552 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [ ]:
%%timeit
option3 = np.cos(values)


1.08 ms ± 263 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


## Ejemplo de módulo: Trabajando con decimales

Los programas de cómputo que ejecutan aritmética con números de punto flotante pueden producir resultados inesperados e imprecisos porque los números de punto flotante no pueden representar adecuadamente todos los número decimales.

In [ ]:
item, rate = 0.70, 1.05
tax = item * rate
total = item + tax
txt, val = ['item','tax','total'], [item,tax,total]  #etiqueta y el valor

for tt, vv in zip(txt, val):
    print(f'{tt:5s} = {vv:.2f}')  # 5 indica que el campo debe tener al menos 5 caracteres de ancho y s es por si tiene menos de 5, se rellene con espacio

item  = 0.70
tax   = 0.73
total = 1.44


Con más decimales

In [ ]:
for tt, vv in zip(txt, val):
    print(f'{tt:8s} = {vv:.20f}')

item     = 0.69999999999999995559
tax      = 0.73499999999999998668
total    = 1.43500000000000005329


Los errores de la aritmética de punto flotante pueden evitarse usando el módulo de Python `decimal`. Este módulo contiene un objeto `Decimal()` con el cual los números de punto flotante pueden representarse con más precisión.

In [ ]:
from decimal import Decimal
item, rate = Decimal('0.70'), Decimal('1.05')
tax = item * rate
total = item + tax

txt, val = ['item','tax','total'], [item,tax,total]

for tt, vv in zip(txt, val):
    print(f'{tt:5s} = {vv:.2f}')

item  = 0.70
tax   = 0.74
total = 1.44


With more decimals

In [ ]:
for tt, vv in zip(txt, val):
    print(f'{tt:5s} = {vv:.20f}')

item  = 0.70000000000000000000
tax   = 0.73500000000000000000
total = 1.43500000000000000000


# Creando un módulo

Los módulo son muy convenientes para almacenar funciones relacionadas en un solo archivo, de manera que podamos mantener el orden en nuestro proyecto y además reutilizar esas funciones en distintos lugares.

Por ejemplo, archivo **formato.py** que está en la misma carpeta que este cuaderno de Jupyter contiene una función llamada `set_css_style`, con la cual controlamos parte del formato de este cuaderno.

In [ ]:
from formato import set_css_style

ModuleNotFoundError: No module named 'formato'

Para crear un módulo, simplemente almacenamos una o más definiciones (de funciones, variables, clases) en un archivo con extensión **.py**. Si el archivo está en la misma carpeta que el archivo de Python en ejecución, lo podemos importar directamente.